Lib/importlib/metadata.py

```diff
@@ -0,0 +1,271 @@
+import os
+import re
+import abc
+import sys
+import email
+import itertools
+import contextlib
+
+from configparser import ConfigParser
+from importlib import import_module
+from importlib.resources import Package
+from pathlib import Path
+from types import ModuleType
+from typing import Callable
+from zipfile import ZipFile
+
+
+__all__ = [
+    'Distribution',
+    'PackageNotFoundError',
+    'distribution',
+    'entry_points',
+    'resolve',
+    'version',
+    ]
+
+
+class PackageNotFoundError(ModuleNotFoundError):
+    """The package's metadata was not found."""
+
+
+class Distribution:
+    """A Python distribution package."""
+
+    @abc.abstractmethod
+    def load_metadata(self, name):
+        """Attempt to load metadata given by the name.
+
+        :param name: The name of the distribution package.
+        :return: The metadata string if found, otherwise None.
+        """
+
+    @classmethod
+    def from_name(cls, name):
+        """Return the Distribution for the given package name.
+
+        :param name: The name of the distribution package to search for.
+        :return: The Distribution instance (or subclass thereof) for the named
+            package, if found.
+        :raises PackageNotFoundError: When the named package's distribution
+            metadata cannot be found.
+        """
+        for resolver in cls._discover_resolvers():
+            resolved = resolver(name)
+            if resolved is not None:
+                return resolved
+        else:
+            raise PackageNotFoundError(name)
+
+    @staticmethod
+    def _discover_resolvers():
+        """Search the meta_path for resolvers."""
+        declared = (
+            getattr(finder, 'find_distribution', None)
+            for finder in sys.meta_path
+            )
+        return filter(None, declared)
+
+    @classmethod
+    def from_module(cls, module):
+        """Discover the Distribution package for a module."""
+        return cls.from_name(cls.name_for_module(module))
+
+    @classmethod
+    def from_named_module(cls, mod_name):
+        return cls.from_module(import_module(mod_name))
+
+    @staticmethod
+    def name_for_module(module):
+        """Given an imported module, infer the distribution package name."""
+        return getattr(module, '__dist_name__', module.__name__)
```

Comment:
Who is setting `__dist_name__`?

Reply:
This behavior has been removed upstream.